src/job_cache/job_cache.cpp

```diff
@@ -215,49 +224,81 @@ FindJobResponse Cache::read(const FindJobRequest &find_request) {
-      // TODO: Try to reconnect to the daemon, launching our own if need be.
-      //       if that fails then depending on user preference either fail
-      //       or return a cache miss.
-      wcl::log::fatal("Cache::read(): failed receiving message");
+      // TODO: Add config var to determine if fail is a cache miss
+      if (false) {
+        return wcl::result_value<FindJobError>(FindJobResponse(wcl::optional<MatchingJob>{}));
+      }
+
+      wcl::log::error("Cache::read(): failed receiving message");
+      return wcl::result_error<FindJobResponse>(FindJobError::FailedMessageReceive);
     }
 
     if (state == MessageParserState::StopSuccess && messages.empty()) {
-      // TODO: Try to reconnect to the daemon, launching our own if need be.
-      //       if that fails then depending on user preference either fail
-      //       or return a cache miss.
-      wcl::log::fatal("Cache::read(): daemon exited without responding");
-    }
+      // TODO: Add config var to determine if fail is a cache miss
+      if (false) {
+        return wcl::result_value<FindJobError>(FindJobResponse(wcl::optional<MatchingJob>{}));
+      }
 
-    // MessageParser tries to avoid this but we should defend against
-    // the case where no error has yet occured but messages is still empty.
-    if (state == MessageParserState::Continue && messages.empty()) {
-      continue;
+      wcl::log::error("Cache::read(): daemon exited without responding");
+      return wcl::result_error<FindJobResponse>(FindJobError::NoResponse);
     }
 
-    if (messages.size() != 1) {
+    if (messages.size() > 1) {
       wcl::log::info("message.size() == %lu", messages.size());
       for (const auto &message : messages) {
         wcl::log::info("message.size() = %lu, message = '%s'", message.size(), message.c_str());
       }
-      wcl::log::fatal("Cache::read(): daemon responded with too many results");
+      wcl::log::error("Cache::read(): daemon responded with too many results");
+      return wcl::result_error<FindJobResponse>(FindJobError::TooManyResponses);
     }
 
-    break;
-  }
+    // We have a singular valid response
+    if (messages.size() == 1) {
+      break;
+    }
+
+    // MessageParser tries to avoid this but we should defend against
+    // the case where no error has yet occured but messages is still empty.
+  } while (state == MessageParserState::Continue);
 
   wcl::log::info("Cache::read(): message rx: %s", messages[0].c_str());
 
   JAST json;
   std::stringstream parseErrors;
   if (!JAST::parse(messages[0], parseErrors, json)) {
-    wcl::log::fatal("Cache::read(): failed to parse daemon response");
+    wcl::log::error("Cache::read(): failed to parse daemon response");
+    return wcl::result_error<FindJobResponse>(FindJobError::FailedParseResponse);
+  }
+
+  return wcl::result_value<FindJobError>(FindJobResponse(json));
+}
+
+FindJobResponse Cache::read(const FindJobRequest &find_request) {
+  for (int i = 0; i < 10; i++) {
```

Comment:
I think we want exponential back off here, but we also want a static variable here that counts the total number of times we've failed to connect so we don't spend too much time in the "no-fail" config case. We can wait to add that once we have an actual no-fail config case though but maybe a TODO here would be good.

This flag will also prevent us from spamming them with errors/warnings when they're in a long run that they want to keep running.